### PR TITLE
Create Unbounded_knapsack.cpp

### PR DESCRIPTION
**Pull request by yeolesanskruti**
#### Description of Change

Issue - #2782 

Added Unbounded Knapsack Algorithm in Dp section with algorithm, task and test cases. Also code is verified using gfg ide.

For reference : 

Screenshot 2024-10-07 213621

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 
Please review and suggest changes if any @realstealthninja

### CODE REVIEW

File: dynamic_programming/Unbounded_knapsack.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/* 
+The Unbounded Knapsack problem allows you to take unlimited quantities of each item. The goal is to maximize the total value
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_knapsack.cpp:2]: please follow the [contribution guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)  to format the header

File: dynamic_programming/Unbounded_knapsack.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+Test Cases -
+Input: 
+N = 4, W = 8, val[] = {6, 1, 7, 7}, wt[] = {1, 3, 4, 5}
+Output:
+48
+*/
+
+// Code
+#include <bits/stdc++.h>
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_knapsack.cpp:19]: dont use bits/stdc++ its not portable.
```suggestion
```

File: dynamic_programming/Unbounded_knapsack.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+Note: Each item can be taken any number of times.
+
+Test Cases -
+Input: 
+N = 4, W = 8, val[] = {6, 1, 7, 7}, wt[] = {1, 3, 4, 5}
+Output:
+48
+*/
+
+// Code
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_knapsack.cpp:18]:
```suggestion
```

File: dynamic_programming/Unbounded_knapsack.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+variation, you can pick any item as many times as needed, as long as the total weight stays within the knapsack's capacity. 
+The problem is commonly solved using dynamic programming.
+
+TASK - Given a set of N items, each with a weight and a value, represented by the array wt and val respectively. Also, a knapsack with weight limit W.
+The task is to fill the knapsack in such a way that we can get the maximum profit. Return the maximum profit.
+Note: Each item can be taken any number of times.
+
+Test Cases -
+Input: 
+N = 4, W = 8, val[] = {6, 1, 7, 7}, wt[] = {1, 3, 4, 5}
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_knapsack.cpp:13]: test cases should be in a separate functions function.

File: dynamic_programming/Unbounded_knapsack.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+Output:
+48
+*/
+
+// Code
+#include <bits/stdc++.h>
+using namespace std;
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_knapsack.cpp:20]:
```suggestion
```
using namespace std is bad practice and should not be used

File: dynamic_programming/Unbounded_knapsack.cpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+int unboundedKnapsack(int N, int W, vector<int> val, vector<int> wt)
+{
+    vector<vector<int>> dp(N,vector<int>(W+1,-1));
+    return KnapSackFilling(N-1,W,val,wt,dp);
+}
+int main() {
+    int N;
+    N=4;
+    vector<int> wt= {1, 3, 4, 5};
+    vector<int> val= {6, 1, 7, 7};
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_knapsack.cpp:47]: since ints are used pass negative numbers as inputs or
use std::uint16_t from <cstdint>

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by yeolesanskruti** [comment]
Sure @realstealthninja I would update the PR soon.
Thank you.

**Comment by yeolesanskruti** [comment]
I have updated the documentation and code accordingly. Please review @realstealthninja .
Thank you.

File: dynamic_programming/Unbounded_knapsack.cpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+ * @brief Unbounded Knapsack problem using Dynamic Programming
+ * @details 
+ * The Unbounded Knapsack problem allows you to take unlimited quantities of each item. The goal is to maximize the total 
+ * value without exceeding the given knapsack capacity. Unlike the 0/1 knapsack, where each item can be taken only once, 
+ * in this variation, you can pick any item as many times as needed, as long as the total weight stays within the knapsack's capacity. 
+ * 
+ * @task Given a set of N items, each with a weight and a value, represented by the array `wt` and `val` respectively, and a knapsack 
+ * with a weight limit W, the task is to fill the knapsack in such a way that the maximum profit is obtained. Return the maximum profit.
+ * 
+ * Note: Each item can be taken any number of times.
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_knapsack.cpp:12]:
```suggestion
 * @note Each item can be taken any number of times.
```
note is a valid tag

File: dynamic_programming/Unbounded_knapsack.cpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    std::vector<std::uint16_t> val = {6, 1, 7, 7}; // Values of the items
+    std::uint16_t W = 8; // Maximum capacity of the knapsack
+
+    // Test the function and assert the expected output
+    assert(unboundedKnapsack(N, W, val, wt) == 48);
+    std::cout << "All test cases passed!" << std::endl;
+}
+
+int main() {
+    test_cases();
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_knapsack.cpp:81]:
```suggestion
    tests(); // run self test implementation
```

File: dynamic_programming/Unbounded_knapsack.cpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    std::uint16_t N = 4; // Number of items
+    std::vector<std::uint16_t> wt = {1, 3, 4, 5}; // Weights of the items
+    std::vector<std::uint16_t> val = {6, 1, 7, 7}; // Values of the items
+    std::uint16_t W = 8; // Maximum capacity of the knapsack
+
+    // Test the function and assert the expected output
+    assert(unboundedKnapsack(N, W, val, wt) == 48);
+    std::cout << "All test cases passed!" << std::endl;
+}
+
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_knapsack.cpp:79]:
```suggestion

/**
 * @brief main function
 * @return 0 on successful exit
 */
```

File: dynamic_programming/Unbounded_knapsack.cpp [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * @brief Function to run test cases for the unbounded knapsack implementation.
+ */
+void test_cases() {
+    std::uint16_t N = 4; // Number of items
+    std::vector<std::uint16_t> wt = {1, 3, 4, 5}; // Weights of the items
+    std::vector<std::uint16_t> val = {6, 1, 7, 7}; // Values of the items
+    std::uint16_t W = 8; // Maximum capacity of the knapsack
+
+    // Test the function and assert the expected output
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_knapsack.cpp:75]: add more test cases please

File: dynamic_programming/Unbounded_knapsack.cpp [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * @brief Wrapper function to initiate the unbounded knapsack calculation.
+ *
+ * @param N Number of items.
+ * @param W Maximum weight capacity of the knapsack.
+ * @param val Vector of values corresponding to the items.
+ * @param wt Vector of weights corresponding to the items.
+ * @return The maximum value that can be obtained for the given capacity.
+ */
+int unboundedKnapsack(std::uint16_t N, std::uint16_t W, const std::vector<std::uint16_t>& val, const std::vector<std::uint16_t>& wt) {
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_knapsack.cpp:61]: does this really return any negative numbers?

File: dynamic_programming/Unbounded_knapsack.cpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    std::vector<std::vector<int>> dp(N, std::vector<int>(W + 1, -1));
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_knapsack.cpp:62]: shouldn't these be unsigned as well?

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: dynamic_programming/Unbounded_knapsack.cpp (hunk 11 above)
Comment (yeolesanskruti) [reply dynamic_programming/Unbounded_knapsack.cpp:61]: No, as the weights of the items in knapsack is positive.

File: dynamic_programming/Unbounded_knapsack.cpp (hunk 12 above)
Comment (yeolesanskruti) [reply dynamic_programming/Unbounded_knapsack.cpp]: No, as DP using memoisation are intialized with -1 values and weight sum is positive to check whether the weight value exists are not -1 is used.

**Comment by yeolesanskruti** [comment]
@realstealthninja  I have updated accordingly. Please review.

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+ * 
+ * @author [Sanskruti Yeole](https://github.com/yeolesanskruti)
+ * @see dynamic_programming/0_1_knapsack.cpp
+ */
+
+#include <iostream>  ///< Standard input-output stream
+#include <vector>   ///< Standard library for using dynamic arrays (vectors)
+#include <cassert>  ///< For using assert function to validate test cases
+#include <cstdint>  ///< For fixed-width integer types like std::uint16_t
+
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_0_1_Knapsack.cpp:31]:
```suggestion
#include <iostream>  // Standard input-output stream
#include <vector>   // Standard library for using dynamic arrays (vectors)
#include <cassert>  // For using assert function to validate test cases
#include <cstdint>  // For fixed-width integer types like std::uint16_t

```

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+ * @param i Current index in the value and weight vectors.
+ * @param W Remaining capacity of the knapsack.
+ * @param val Vector of values corresponding to the items.
+ * @param wt Vector of weights corresponding to the items.
+ * @param dp 2D vector for memoization to avoid redundant calculations.
+ * @return The maximum value that can be obtained for the given index and capacity.
+ */
+int KnapSackFilling(std::uint16_t i, std::uint16_t W, 
+                    const std::vector<std::uint16_t>& val, 
+                    const std::vector<std::uint16_t>& wt,
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_0_1_Knapsack.cpp:57]: Couldn't values and weights be a std::pair?

That is
```cpp
std::vector<std::pair<std::uint16_t, std::uint16_t> data
```

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    // Test the function and assert the expected output
+    assert(unboundedKnapsack(N3, W3, val3, wt3) == 27);
+    std::cout << "Maximum Knapsack value " << unboundedKnapsack(N3, W3, val3, wt3) << std::endl;
+  
+    std::cout << "All test cases passed!" << std::endl;
+
+}
+
+int main() {
+    tests(); // Run test cases to validate implementation
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_0_1_Knapsack.cpp:131]:
```suggestion
    tests(); // Run self test implementation 
```

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+    std::vector<std::uint16_t> val3 = {5, 11, 13};// Values of the items 
+    std::uint16_t W3 = 27;// Maximum capacity of the knapsack 
+    // Test the function and assert the expected output
+    assert(unboundedKnapsack(N3, W3, val3, wt3) == 27);
+    std::cout << "Maximum Knapsack value " << unboundedKnapsack(N3, W3, val3, wt3) << std::endl;
+  
+    std::cout << "All test cases passed!" << std::endl;
+
+}
+
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_0_1_Knapsack.cpp:129]:
```suggestion

/**
 * @brief main function
 * @return 0 on successful exit
 */
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp (hunk 14 above)
Comment (yeolesanskruti) [reply dynamic_programming/Unbounded_0_1_Knapsack.cpp:57]: We can pair it, but the input of the algorithm is taken like this and it would become complex as well.

**Comment by yeolesanskruti** [comment]
@realstealthninja  I have updated the file accordingly. I hope so it is fit for merging now.

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+static void tests() {
+    // Test Case 1
+    std::uint16_t N1 = 4;  // Number of items
+    std::vector<std::uint16_t> wt1 = {1, 3, 4, 5}; // Weights of the items
+    std::vector<std::uint16_t> val1 = {6, 1, 7, 7}; // Values of the items
+    std::uint16_t W1 = 8; // Maximum capacity of the knapsack
+    // Test the function and assert the expected output
+    assert(unboundedKnapsack(N1, W1, val1, wt1) == 48);
+    std::cout << "Maximum Knapsack value " << unboundedKnapsack(N1, W1, val1, wt1) << std::endl;
+
```
[the remark this answers is not in the document]
Comment (yeolesanskruti) [reply dynamic_programming/Unbounded_0_1_Knapsack.cpp:108]: Done.

**Comment by yeolesanskruti** [comment]
@realstealthninja done all changes accordingly.

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+ *        an unbounded knapsack approach.
+ *
+ * @param i Current index in the value and weight vectors.
+ * @param W Remaining capacity of the knapsack.
+ * @param val Vector of values corresponding to the items.
+ * @param wt Vector of weights corresponding to the items.
+ * @param dp 2D vector for memoization to avoid redundant calculations.
+ * @return The maximum value that can be obtained for the given index and capacity.
+ */
+int KnapSackFilling(std::uint16_t i, std::uint16_t W,
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_0_1_Knapsack.cpp:55]: this type could be too small lets switch to uint32_t

Comment (realstealthninja) [inline dynamic_programming/Unbounded_0_1_Knapsack.cpp:55]: if the maximum value cannot be negative better to use uint32_t or if its an index size_t

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+ * @brief main function
+ * @return 0 on successful exit
+ */
+int main() {
+    tests(); // Run self test implementation 
+    return 0;
+}
+
+}
+}
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_0_1_Knapsack.cpp:149]: name spaces should not include main & test functions. they should have comments specifying which name space they enclose
```cpp
} // namespace_name
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp (hunk 18 above)
Comment (yeolesanskruti) [reply dynamic_programming/Unbounded_0_1_Knapsack.cpp:55]: @realstealthninja we have not specified any conditions on value and weight array and the algorithm is important which is explained well, one can do required changes accordingly.

**Comment by yeolesanskruti** [comment]
@realstealthninja  all changes are done accordingly. Hope so code is ready for merge

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp (hunk 19 above)
Comment (yeolesanskruti) [reply dynamic_programming/Unbounded_0_1_Knapsack.cpp]: done

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp (hunk 18 above)
Comment (realstealthninja) [reply dynamic_programming/Unbounded_0_1_Knapsack.cpp:55]: please comment on this

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by yeolesanskruti** [comment]
@realstealthninja  done.  Hope that file is ready for the merge now.

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp (hunk 18 above)
Comment (realstealthninja) [reply dynamic_programming/Unbounded_0_1_Knapsack.cpp:57]: ?? why is the return type still an int? is there a possibility the maximum value is negative?

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by yeolesanskruti** [comment]
@realstealthninja The return type is int because the knapsack value is typically non-negative.

**Comment by realstealthninja** [comment]
> @realstealthninja No, in the standard unbounded knapsack problem, the maximum value cannot be negative. The problem is designed to maximize the total value of items placed in the knapsack, and item values are typically non-negative. If all items have zero or negative values, the optimal solution would be to select no items, resulting in a maximum value of 0. Thus, the maximum value is always non-negative.

oh i see! thanks for the clarification

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+
+} // unbounded_knapsack
+
+} // dynamic_programming
+
+/**
+ * @brief Function to run test cases for the unbounded knapsack implementation.
+ * 
+ * @details This function tests the implementation with predefined test cases.
+ * It asserts the expected outcomes to ensure correctness of the algorithm.
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_0_1_Knapsack.cpp:103]:
```suggestion
 * @brief self test implementation
 * @return void
```

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+    assert(unboundedKnapsack(N1, W1, val1, wt1) == 48);
+    std::cout << "Maximum Knapsack value " << unboundedKnapsack(N1, W1, val1, wt1) << std::endl;
+
+    // Test Case 2
+    std::uint16_t N2 = 3; // Number of items
+    std::vector<std::uint16_t> wt2 = {10, 20, 30}; // Weights of the items
+    std::vector<std::uint16_t> val2 = {60, 100, 120}; // Values of the items
+    std::uint16_t W2 = 5; // Maximum capacity of the knapsack
+    // Test the function and assert the expected output
+    assert(unboundedKnapsack(N2, W2, val2, wt2) == 0);
```
Comment (realstealthninja) [inline dynamic_programming/Unbounded_0_1_Knapsack.cpp:121]: add a test case where this returns negative values since return type is negative

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by yeolesanskruti** [comment]
@realstealthninja According to the problem constraints, both the weights and values of items are strictly greater than or equal to zero, which ensures the result will always be non-negative. Therefore, adding a test case that returns a negative value is invalid for this problem. If you're seeing negative values in your output, it likely indicates an issue with the implementation or the problem definition itself.

File: dynamic_programming/Unbounded_0_1_Knapsack.cpp (hunk 21 above)
Comment (yeolesanskruti) [reply dynamic_programming/Unbounded_0_1_Knapsack.cpp:119]: @realstealthninja According to the problem constraints, both the weights and values of items are strictly greater than zero, which ensures the result will always be non-negative. Therefore, adding a test case that returns a negative value is invalid for this problem. If you're seeing negative values in your output, it likely indicates an issue with the implementation or the problem definition itself.

Comment (realstealthninja) [reply dynamic_programming/Unbounded_0_1_Knapsack.cpp:119]: > @realstealthninja According to the problem constraints, both the weights and values of items are strictly greater than or equal to zero, which ensures the result will always be non-negative. Therefore, adding a test case that returns a negative value is invalid for this problem. If you're seeing negative values in your output, it likely indicates an issue with the implementation or the problem definition itself.

Therefore the return value of the function should not be an int. it should be an unsigned integer.

Comment (yeolesanskruti) [reply dynamic_programming/Unbounded_0_1_Knapsack.cpp:119]: Sure @realstealthninja, changes are done accordingly.

**Review by realstealthninja — APPROVED** [review]
LGTM thanks for this :heart:

**Comment by Pranjal-231003** [comment]
Allow me to work on this , I am kinda late to hacktober so want to complete my PRs
